editor/src/clj/editor/code/view.clj

```diff
@@ -1025,6 +1027,240 @@
+    :hover (if (:hoverable region)
+             :hover-showing-lsp-regions
+             :hover-cursor-lsp-regions)
+    :regions))
+
+(defn- set-properties
+  "Return transaction steps for the view changes"
+  [view-node undo-grouping values-by-prop-kw]
+  (let [resource-node (g/node-value view-node :resource-node)
+        resource-node-type (g/node-type* resource-node)]
+    (into (prelude-tx-data view-node undo-grouping values-by-prop-kw)
+          (mapcat (fn [[prop-kw value]]
+                    (case prop-kw
+                      :cursor-ranges
+                      (if (g/has-property? resource-node-type :cursor-ranges)
+                        (g/set-property resource-node :cursor-ranges value)
+                        (g/set-property view-node :fallback-cursor-ranges value))
+
+                      :regions
+                      (let [{:keys [diagnostics hover-showing-lsp-regions hover-cursor-lsp-regions regions]}
+                            (group-by region->prop-kw value)]
+                        (concat
+                          (g/set-property view-node :hover-showing-lsp-regions hover-showing-lsp-regions)
+                          (g/set-property view-node :hover-cursor-lsp-regions hover-cursor-lsp-regions)
+                          (g/set-property view-node :diagnostics (or diagnostics []))
+                          (g/set-property resource-node prop-kw (or regions []))))
+
+                      ;; Several actions might have invalidated rows since
+                      ;; we last produced syntax-info. We keep an ever-
+                      ;; growing history of invalidated-rows. Then when
+                      ;; producing syntax-info we find the first invalidated
+                      ;; row by comparing the history of invalidated rows to
+                      ;; what it was at the time of the last call. See the
+                      ;; invalidated-row function for details.
+                      :invalidated-row
+                      (g/update-property resource-node :invalidated-rows conj value)
+
+                      ;; The :indent-type output in the resource node is
+                      ;; cached, but reads from disk unless a value exists
+                      ;; for the :modified-indent-type property.
+                      :indent-type
+                      (g/set-property resource-node :modified-indent-type value)
+
+                      ;; The :lines output in the resource node is uncached.
+                      ;; It reads from disk unless a value exists for the
+                      ;; :modified-lines property. This means only modified
+                      ;; or currently open files are kept in memory.
+                      :lines
+                      (g/set-property resource-node :modified-lines value)
+
+                      ;; All other properties are set on the view node.
+                      (g/set-property view-node prop-kw value))))
+          values-by-prop-kw)))
```

Comment:
Minor refactoring: split `set-properties!` into `set-properties!` (do the tx) and `set-properties` (return tx steps)